### DISCUSSION TITLE
Help With Decorator Parameter Hints

### DISCUSSION THREAD

**Discussion by Logon27** [microsoft/pylance-release#4839] [Q&A] [answered]
Hello. I am manually using a python decorator to wrap a function. Inside the decorator I have proper `functools.wraps` calls for updating the metadata. However, Pylance gives me this terrible "_Wrapped" syntax...
![image](https://github.com/microsoft/pylance-release/assets/32345277/2baa5899-4565-476b-9511-20c5ea31dc55)
When it should be shown like...
![image](https://github.com/microsoft/pylance-release/assets/32345277/7c72ad40-d1bd-41ab-ac14-7620a7671698)
Is there any way to get Pylance to display the correct function parameter hints? I have tried a bunch of different methods and nothing seems to work.

**Comment by erictraut** [comment]
Could you provide a minimal, self-contained code sample (in text form) that shows the problem you're trying to solve? It's difficult to answer your question from partial screen shots.

**Reply by Logon27** [reply]
Sure @erictraut, sorry if I was not clear. Take for example this dummy code snippet...
```python
from typing import Tuple
import functools

def debug_decorator(dense_debug):

    @functools.wraps(dense_debug)
    def Dense(*args, **kwargs):
        init_fun_debug, apply_fun_debug = dense_debug(*args, **kwargs)

        @functools.wraps(init_fun_debug)
        def init_fun(input_shape: Tuple):
            return input_shape
            
        @functools.wraps(apply_fun_debug)
        def apply_fun(inputs: Tuple):
            return inputs
        
    return Dense

@debug_decorator
def Dense(out_dim: int):

    def init_fun(input_shape: Tuple):
        return input_shape

    # kwargs is necessary due to rng being passed to some apply functions.
    def apply_fun(inputs: Tuple):
        return inputs

    return init_fun, apply_fun
```
If you call the function directly with the decorator annotation on it. You get nice parameter hints...
![image](https://github.com/microsoft/pylance-release/assets/32345277/9cb4e564-e029-46a0-ba65-982a4f7349bb)

But if you remove the `@debug_decorator` annotation and you directly call the decorator to wrap the function. You get this really ugly and useless wrapped syntax...
![image](https://github.com/microsoft/pylance-release/assets/32345277/6fc9b3dd-a2c1-4d9f-ac7d-c9aabf2151f2)

Hopefully I am clear enough about what my issue is now. If not then please let me know.

**Reply by erictraut** [reply]
I'm not sure I fully understand what this code is supposed to do. Is it possible that you omitted a return statement in the first `Dense` function?

I've taken a stab at annotating the code, but I needed to make some assumptions. Let me know if my assumptions were incorrect.

```python
from typing import Any, Callable, ParamSpec
import functools

P1 = ParamSpec("P1")

def debug_decorator(
    dense_debug: Callable[P1, tuple[Callable[..., Any], Callable[..., Any]]]
):
    @functools.wraps(dense_debug)
    def dense(*args: P1.args, **kwargs: P1.kwargs):
        init_fun_debug, apply_fun_debug = dense_debug(*args, **kwargs)

        @functools.wraps(init_fun_debug)
        def init_fun(input_shape: tuple[int, ...]):
            return input_shape

        @functools.wraps(apply_fun_debug)
        def apply_fun(inputs: tuple[int, ...]):
            return inputs

        return init_fun, apply_fun

    return dense

def dense(out_dim: int):
    def init_fun(input_shape: tuple[int, ...]):
        return input_shape

    def apply_fun(inputs: tuple[int, ...]):
        return inputs

    return init_fun, apply_fun

dense2 = debug_decorator(dense)
```

**Reply by Logon27** [reply]
That example was me cutting all out dozens of lines of code to give a dummy example. Let's just simplify to a trivial example so we are on the same page here. An example function that just takes in a number and returns it. An example_decorator that just wraps the function and prints the value.
```python
import functools

def example_decorator(example_debug):

    @functools.wraps(example_debug)
    def new_example(*args, **kwargs):
        print(f"my number: {args[0]}")
        example_debug(*args, **kwargs)

    return new_example

def example(my_number: int):
    return my_number

example2 = example_decorator(example)

```
![image](https://github.com/microsoft/pylance-release/assets/32345277/63954139-08c7-4d39-a84c-cb006c006801)
Edit: I changed the inner decorator function to something unique just so we are confusing the function scope here.
Edit2: Whoops. Removed the decorator annotation since we are calling the decorator directly.

**Reply by erictraut** [reply]
You can use a `ParamSpec` to preserve the signature of the decorated function:

```python
P = ParamSpec("P")

def example_decorator(example_debug: Callable[P, Any]):
    @functools.wraps(example_debug)
    def new_example(*args: P.args, **kwargs: P.kwargs):
        print(f"my number: {args[0]}")
        example_debug(*args, **kwargs)

    return new_example
```

**Reply by erictraut** [reply, marked as answer]
If you want to preserve the return type as well, you'll need to return it from within your `new_example` inner function.

```python
P = ParamSpec("P")
R = TypeVar("R")

def example_decorator(example_debug: Callable[P, R]) -> Callable[P, R]:
    @functools.wraps(example_debug)
    def new_example(*args: P.args, **kwargs: P.kwargs) -> R:
        print(f"my number: {args[0]}")
        return example_debug(*args, **kwargs)

    return new_example
```

**Reply by erictraut** [reply]
For more details about `ParamSpec`, refer to [PEP 612](https://peps.python.org/pep-0612/).

**Reply by Logon27** [reply]
The new parameter hint...
![image](https://github.com/microsoft/pylance-release/assets/32345277/a5f87b8e-e2fa-4b3e-ab4f-47a0496924bf)

Like yes, the signature has improved slightly. But it is still a far cry from the original parameter hint...
![image](https://github.com/microsoft/pylance-release/assets/32345277/31fe322c-e9c8-4794-b0af-73436b8e95c4)

For any non-trivial example, this "_Wrapped" syntax is still pretty hard to read and does not respect new line formatting and color differentiation. Is there anything I can do to improve this even more?

**Reply by Logon27** [reply]
> ```python
> P = ParamSpec("P")
> R = TypeVar("R")
> 
> def example_decorator(example_debug: Callable[P, R]) -> Callable[P, R]:
>     @functools.wraps(example_debug)
>     def new_example(*args: P.args, **kwargs: P.kwargs) -> R:
>         print(f"my number: {args[0]}")
>         return example_debug(*args, **kwargs)
> 
>     return new_example
> ```

@erictraut Bizarre, only your second example with both the ParamSpec and TypeVar produces the correct parameter hint. Your first example doesn't give the same level of parameter hints. Your second example (which works)...
![image](https://github.com/microsoft/pylance-release/assets/32345277/09732692-2833-46d5-bcff-b322ff3895bb)
versus your first example snippet...
![image](https://github.com/microsoft/pylance-release/assets/32345277/a5f87b8e-e2fa-4b3e-ab4f-47a0496924bf)

**Reply by Logon27** [reply]
I have to get off for the night. Thank you Eric, I appreciate all the help. I know nothing about these parameter specs so I will have to read up on it. A tad annoying that I have to specifically type hint my code with this format in order to get the proper parameter hints. But it is better than having no working solution at all.

**Reply by erictraut** [reply]
`_Wrapped` comes from the `functools.wraps` call. `_Wrapped` is an object that can be called but also supports several attributes including `__name__` and `__qualname__`. You can view the definition by right-clicking on `wraps` and choosing "Go To Declaration".

**Reply by Logon27** [reply]
That I did know. I just don't really understand why I have to give these special type declarations for proper parameter hinting. So I will have to read up on it as you suggested. As for my use case. I use decorators pretty heavily for debugging my deep learning functions. It helps me separate the core logic from any debug logic. And not being able to see proper parameter hints with so many optional parameters was bugging me extensively. Thanks again for the help Eric.

For anyone else that has a similar issue. Here is an example of how I ended up implementing the solution. It should work for python 3.7 and up. `Callable[P, R]` is needed for both the parameters and return type of the decorator. And both `P.args` and `P.kwargs` must be present, as well as the `R` return value. If you do not use all these type hints then you will not get the clean parameter hints you expect with all the nice formatting.

```python
import sys
if sys.version_info < (3, 10):
    from typing_extensions import ParamSpec
else:
    from typing import ParamSpec
from typing import Callable, TypeVar

P = ParamSpec("P")
R = TypeVar("R")

def example_decorator(example_debug: Callable[P, R]) -> Callable[P, R]:
    @functools.wraps(example_debug)
    def new_example(*args: P.args, **kwargs: P.kwargs) -> R:
        print(f"my number: {args[0]}")
        return example_debug(*args, **kwargs)

    return new_example

def example(my_number: int):
    return my_number

example2 = example_decorator(example)
```
